moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,548 @@
+   *   writer.beginArray();
+   *   writer.value(1);
+   *   integersAdapter.toJson(writer, Arrays.asList(2, 3, 4));
+   *   writer.value(5);
+   *   writer.endArray();
+   * }
+   * ```
+   *
+   *
+   * With flattening we can create JSON with a single array `[1,2,3,4,5]`:
+   *
+   * ```java
+   * JsonAdapter<List<Integer>> integersAdapter = ...
+   *
+   * public void writeNumbers(JsonWriter writer) {
+   *   writer.beginArray();
+   *   int token = writer.beginFlatten();
+   *   writer.value(1);
+   *   integersAdapter.toJson(writer, Arrays.asList(2, 3, 4));
+   *   writer.value(5);
+   *   writer.endFlatten(token);
+   *   writer.endArray();
+   * }
+   * ```
+   *
+   *
+   * This method flattens arrays within arrays:
+   *
+   * Emit:       `[1, [2, 3, 4], 5]`
+   * To produce: `[1, 2, 3, 4, 5]`
+   *
+   * It also flattens objects within objects. Do not call [name] before writing a flattened
+   * object.
+   *
+   * Emit:       `{"a": 1, {"b": 2}, "c": 3}`
+   * To Produce: `{"a": 1, "b": 2, "c": 3}`
+   *
+   *
+   * Other combinations are permitted but do not perform flattening. For example, objects inside of
+   * arrays are not flattened:
+   *
+   * Emit:      ` [1, {"b": 2}, 3, [4, 5], 6]`
+   * To Produce: `[1, {"b": 2}, 3, 4, 5, 6]`
+   *
+   *
+   * This method returns an opaque token. Callers must match all calls to this method with a call
+   * to [endFlatten] with the matching token.
+   */
+  @CheckReturnValue
+  public fun beginFlatten(): Int {
+    val context = peekScope()
+    check(!(context != JsonScope.NONEMPTY_OBJECT && context != JsonScope.EMPTY_OBJECT && context != JsonScope.NONEMPTY_ARRAY && context != JsonScope.EMPTY_ARRAY)) { "Nesting problem." }
+    val token = flattenStackSize
+    flattenStackSize = stackSize
+    return token
+  }
+
+  /** Ends nested call flattening created by [beginFlatten].  */
+  public fun endFlatten(token: Int) {
+    flattenStackSize = token
+  }
+
+  /**
+   * Returns a [JsonPath](http://goessner.net/articles/JsonPath/) to the current location
+   * in the JSON value.
+   */
+  @get:CheckReturnValue
+  public val path: String
+    get() = JsonScope.getPath(stackSize, scopes, pathNames, pathIndices)
+
+  /** Returns the tag value for the given class key.  */
+  @CheckReturnValue
+  public fun <T> tag(clazz: Class<T>): T? {
+    @Suppress("UNCHECKED_CAST")
+    return if (tags == null) null else tags!![clazz] as T?
+  }
+
+  /** Assigns the tag value using the given class key and value.  */
+  public fun <T> setTag(clazz: Class<T>, value: T) {
+    require(clazz.isAssignableFrom(value!!::class.java)) { "Tag value must be of type " + clazz.name }
+    if (tags == null) {
+      tags = LinkedHashMap()
+    }
+    tags!![clazz] = value
+  }
+
+  public companion object {
+    /** Returns a new instance that writes UTF-8 encoded JSON to `sink`.  */
+    @JvmStatic
+    @CheckReturnValue
+    public fun of(sink: BufferedSink?): JsonWriter = JsonUtf8Writer(sink)
```

Comment:
Why the nullable here?

Reply:
```java
  JsonUtf8Writer(BufferedSink sink) {
    if (sink == null) {
      throw new NullPointerException("sink == null");
    }
    this.sink = sink;
    pushScope(EMPTY_DOCUMENT);
  }
```

Reply:
that doesn't really answer the question and the IDE even lints your code here for doing it. I recommend trying to better understand the code you are changing here, pasting back code snippets and IDE error messages is not demonstrating understanding

Reply:
Sorry for misread that, it's midnight where I live, sleepy...

Reply:
ea6ef3f63b129ef407c84bc607a9ededd415d84f